samples: boards: stm32 power_mgt stdby and shutdown

FRASTM opened the pull request:
Polarity low is wrong : wakeup pin 2 is active low
so the wakeup must be detected on the rising edge
This is the default wakeup pin2 config.
Apply the recommended sequence to enter low power standby

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/79451

FRASTM committed:
samples: boards: stm32 power_mgt stdby and shutdown

Polarity low is wrong : wakeup pin 2 is active low
so the wakeup must be detected on the rising edge
This is the default wakeup pin2 config.
Apply the recommended sequence to enter low power standby

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: boards: stm32 power standby detect sw reset cause

After flashing the board, the reset pin and the software
reset are detected as reset cause by the device
Put this in the sample to match the rule

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Running the samples/boards/st/power_mgmt/standby_shutdown/ on the samples/boards/st/power_mgmt/standby_shutdown/
```
*** Booting Zephyr OS build v3.7.0-4029-g9fa64e2d3a39 ***

Reset cause: Reset soft

Device ready: disco_l475_iot1


Press and hold the user button:
  when LED2 is OFF to power off
  when LED2 is ON to enter to Standby Mode
```

Is not possible automatically as the test is waiting for user button pressed

erwango commented on samples/boards/st/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -42,16 +42,19 @@ static struct gpio_callback button_cb_data;
-	LL_PWR_EnablePUPDCfg();
+	/*
+	 * Configure wake-up features : WKUP2(PC13) only
+	 *
+	 * The Following Wakeup sequence is highly recommended prior to each Standby mode entry
+	 * mainly when using more than one wakeup source this is to not miss any wakeup event:
+	 * - Disable all used wakeup sources,
+	 * - Clear all related wakeup flags,
+	 * - Re-enable all used wakeup sources,
+	 * - Enter the Standby mode.
+	 */
+	LL_PWR_DisableWakeUpPin(LL_PWR_WAKEUP_PIN2);
+	LL_PWR_ClearFlag_WU(); /* Clear wakeup flags */
 	LL_PWR_EnableWakeUpPin(LL_PWR_WAKEUP_PIN2);
-	/* Clear wakeup flags */
-	LL_PWR_ClearFlag_WU();
 }
```
ok, but what would be the cost of updating using the new DT way to configure wake up pins (https://github.com/zephyrproject-rtos/zephyr/pull/73058)

FRASTM replied:
When using the new DT way to configure wake up pins, 

- a button press for the standby mode immediately wakes up 
- a button press for the power off  mode goes  to power off until button is released (expected behavior)

FRASTM committed:
samples: boards: power_mgmt use wakeup pins of the stm32l4x boards

This PR is changing the method to use the user button
to be the wakeup source to exit from lower power mode

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
sample tested with disco_l475_iot1

FRASTM committed:
dts: arm: st: stm32l4xx serie has 5 wakeup pins

Add the wakeUp pins selection on its GPIO
for all the stm32l4xx devices
Change if needed (see the corresponding datasheet/RefMan)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

JarmouniA commented on samples/boards/st/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -27,33 +28,24 @@
-	/* WKUP2(PC13) only , - active low, pull-up */
-	/* Set pull-ups for standby modes */
-	LL_PWR_EnableGPIOPullUp(LL_PWR_GPIO_C, LL_PWR_GPIO_BIT_13);
-	LL_PWR_IsWakeUpPinPolarityLow(LL_PWR_WAKEUP_PIN2);
```
I don't think setting wake-up pin polarity to low is incorrect, since it's the push-button that determines it, & in this case it's `Active low`.

FRASTM replied:
this is done through the WakeUp pin config, isn't it

JarmouniA replied:
Yes, but only when `wkup-pins-pol` property is present in pwr node, which was removed in board overlay.
The presence of this property result in the definition of the `PWR_STM32_WKUP_PINS_POLARITY` flag
https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/soc/st/stm32/common/stm32_wkup_pins.c#L35
which guards the polarity cfg code
https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/soc/st/stm32/common/stm32_wkup_pins.c#L196

JarmouniA commented on samples/boards/st/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -13,6 +13,7 @@
 #include <stm32_ll_pwr.h>
+#include <zephyr/dt-bindings/gpio/stm32-gpio.h>
 
 #if !defined(CONFIG_SOC_SERIES_STM32L4X)
```
This check is no longer useful.

FRASTM replied:
removed

JarmouniA commented on samples/boards/st/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -27,33 +28,24 @@
 #error "Unsupported board: sw0 devicetree alias is not defined"
```
`SW0_NODE` is no longer used.

FRASTM replied:
removed

FRASTM commented on samples/boards/st/power_mgmt/standby_shutdown/README.rst:
```diff
@@ -7,13 +7,15 @@
+- when LED2 is ON to enter to Standby Mode
+The push button is now being configured as WakeUp Source (WakeUp pin)
+and system enters in the selected low power mode.
+User can release the push button to exit from shutdown mode or from shutdown mode.
```
updated

FRASTM commented on samples/boards/st/power_mgmt/standby_shutdown/README.rst:
```diff
@@ -7,13 +7,15 @@
+User push-button is configured in gpio with interrupt and callback in a basic blinking LED loop.
+User can select the low power mode by pressing and holding the push-button:
+- when LED2 is OFF to enter to Shutdown Mode
+- when LED2 is ON to enter to Standby Mode
```
updated

FRASTM commented on samples/boards/st/power_mgmt/standby_shutdown/README.rst:
```diff
@@ -7,13 +7,15 @@
+with WakeUp pins.
+User push-button is configured in gpio with interrupt and callback in a basic blinking LED loop.
+User can select the low power mode by pressing and holding the push-button:
+- when LED2 is OFF to enter to Shutdown Mode
```
updated

FRASTM commented on samples/boards/st/power_mgmt/standby_shutdown/boards/disco_l475_iot1.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_STM32_WKUP_PINS=y
```
moved to prj.conf

erwango commented on dts/arm/st/l4/stm32l471.dtsi:
```diff
@@ -282,3 +282,25 @@
 	};
 };
+
+&pwr {
+	wkup-pin@1 {
+		wkup-gpios = <&gpioa 0 STM32_PWR_WKUP_PIN_SRC_0>;
+	};
+
+	wkup-pin@2 {
+		wkup-gpios = <&gpioc 13 STM32_PWR_WKUP_PIN_SRC_0>;
+	};
+
+	wkup-pin@3 {
+		wkup-gpios = <&gpioe 6 STM32_PWR_WKUP_PIN_SRC_0>;
+	};
+
+	wkup-pin@4 {
+		wkup-gpios = <&gpioa 2 STM32_PWR_WKUP_PIN_SRC_0>;
+	};
+
+	wkup-pin@5 {
+		wkup-gpios = <&gpioc 5 STM32_PWR_WKUP_PIN_SRC_0>;
+	};
+};
```
Same list is present in `dts/arm/st/l4/stm32l4r5Xi.dtsi`, and after a quick look to https://github.com/STMicroelectronics/STM32_open_pin_data, it seems that those pins are the always the same on all SoC of L4 series, the only gotcha is that on some variants (small variants or packages) some pins are not present.
For instance wkup-pin 3 not available in STM32L471R(E-G).

So my proposal would be to put the whole list in https://github.com/zephyrproject-rtos/zephyr/blame/main/dts/arm/st/l4/stm32l4.dtsi#L481-L482
and then delete nodes if needed in impacted packages.

For instance:
```
&pwr {
	/delete-node/ wkup-pin@3;
};
```

Note: it might require to add nodelabels to allow the use of /delete-node/ syntax:
```
	wkuppin4: wkup-pin@4 { ....}
[...]
&pwr {
	/delete-node/ wkuppin4;
};
```

FRASTM replied:
Yes, the stm32L4 serie has 5 WKUP pins, same for the stm32L47x/8x/9x/ax and stl32L4plus serie, 
but the gpio associated with each wakeup pin (example GPIOC 13 for the wakeUp pin #2) is defined by each stm32L4 datasheet.

JarmouniA commented:
>> I don't think setting wake-up pin polarity to low is incorrect, since it's the push-button that determines it, & in this case it's `Active low`.

> this is done through the WakeUp pin config, isn't it

@FRASTM Yes, but only when `wkup-pins-pol` property is present in pwr node, which was removed in board overlay.
The presence of this property result in the definition of the `PWR_STM32_WKUP_PINS_POLARITY` flag
https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/soc/st/stm32/common/stm32_wkup_pins.c#L35
which guards the polarity cfg code
https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/soc/st/stm32/common/stm32_wkup_pins.c#L196

FRASTM commented:
> @FRASTM Yes, but only when `wkup-pins-pol` property is present in pwr node, which was removed in board overlay. The presence of this property result in the definition of the `PWR_STM32_WKUP_PINS_POLARITY` flag

@JarmouniA , I don not catch, Could you please detail your request

kartben commented:
Ping @JarmouniA

FRASTM commented:
-->  the https://github.com/zephyrproject-rtos/zephyr/issues/79451 is closed
